Lazy Tensor to `torch` MLIR

ramiro050 opened the pull request:
This PR implements a way of compiling a lazy computation captured by the [Lazy Tensor Core](https://github.com/pytorch/pytorch/tree/lazy_tensor_staging/lazy_tensor_core) (LTC) of Pytorch into the `torch` MLIR dialect. The following is a high level description of how this compilation is done:

1. Use the built-in TorchScript plugin of LTC to convert the lazy computation captured by LTC into a `torch.jit.ScriptFunction`
2. Use the `ModuleBuilder.import_function` to convert the `ScriptFunction` into an `MLIR` module
3. Use the the new pass for `ScriptFunction`s in the `torch_mlir.passmanager` to compile the MLIR module all the way down

## Request for feedback

- I am currently modifying the pass pipelines of `torch-mlir` (see [commit](https://github.com/ramiro050/mlir-npcomp/commit/cb1fac76c969cfd6adaefca6382f05db5b84a43d)). The reason this is needed is because the `GlobalizedModule` pass expects a `torch.nn.Module` to be part of the MLIR, which is always true for `torch.jit.ScriptModule` objects, but not for `torch.jit.ScriptFunction` objects. Thus, I decided to turn the existing TorchScript pipeline into two pipelines, one for `ScriptFunction`s and one for `ScriptModule`s. The way I chose to change things works for compiling `torch.jit.ScriptFunction`s, but I am not sure if it is completely correct or if there is a better way of doing it. I would love any feedback on this change.

## Under construction

- `utils/annotator.py`: This file is currently a work in progress. The next step is to change the `torchfx` code so that it uses a similar interface to the one `lazytensor` is using. In particular, the `Builder` used to turn `torch.fx.ModuleGraph` into MLIR will in the future take an extra argument for an `Annotation` object, and the `annotate_forward_args` function in `annotate.py` will be removed
- I currently have to convert the lazy computation to a string then parse it back in to get a TorchScript IR graph. This should be easily fixable by modifying Pytorch, which I plan on doing in the future

silvasean commented on examples/lazytensor_tanh.py:
```diff
@@ -0,0 +1,83 @@
+    pm = PassManager.parse('torchscript-function-to-linalg-on-tensors-backend-pipeline')
+pm.run(mlir_module)
+
+print("BEFORE NPCOMP BACKEND PIPELINE")
```
nit: remove the word "npcomp" (we are avoiding using that going forward)

silvasean commented on examples/torchfx/builder.py:
```diff
@@ -10,8 +10,9 @@
 
-This file is licensed under a pytorch-style license
-See frontends/pytorch/LICENSE for license information.
+Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
nit: would recommend keeping the license in `#` style comments at the top, followed by the docstring.

(thanks for the docstrings btw :) )

silvasean commented on examples/lazytensor_tanh.py:
```diff
@@ -0,0 +1,83 @@
+y = torch.randn(shape, device=device, dtype=dtype)
+
+# Capture lazy computation and convert to TorchScript IR
+graph_str = ltc._LAZYC._get_ltc_tensors_backend([x.tanh() * y.tanh()])
```
what determines the order of arguments to the TorchScript function that gets created? It seems like nothing here knows whether we want `x` or `y` to be the first argument.

ramiro050 replied:
I am not completely sure. From experimentation, the evaluation order does not seem to be from left to right all the time. For example, `x.tanh() * y` captures `y` as the first argument. I will look into ways of controlling the order

silvasean reviewed (approved):
great work!

shabalind commented:
It's great to see LazyTensor+MLIR being connected together! 👏

ramiro050 commented:
## PR Update

- [x] Update `README`
- [x] Change license style to `#` comments
- [x] Change description of `torchscript-function` pipeline
- [x] Remove mentions of `npcomp`
- [x] Added more documentation
- [x] Fixed a few bugs
- [x] Minor style fixes

ramiro050 commented:
Forgot to update e2e tests to use the new backend name: `torchscript-module-to...`. I'll fix it soon

ramiro050 commented:
Done!
